### PR TITLE
Train Faster R-CNN with negative samples

### PR DESCRIPTION
**Pull request by mnc537**
Hello,

This PR allows FasterRCNN to train with negative samples.
**Related to:** #1598 

When [defining the dataset](https://pytorch.org/tutorials/intermediate/torchvision_tutorial.html), one needs to set the field `boxes` in the `target` dict as `torch.zeros((0, 4), dtype=torch.float32)` for negative images since `boxes` is required. 

This is how `target` should look like:
``` 
target = {}
target["boxes"] = torch.zeros((0, 4), dtype=torch.float32)
target["labels"] = torch.zeros((1, 1), dtype=torch.int64)
target["image_id"] = image_id
target["area"] = (boxes[:, 3] - boxes[:, 1]) * (boxes[:, 2] - boxes[:, 0])
target["iscrowd"] = torch.zeros((0,), dtype=torch.int64)
```

### CODE REVIEW

**Comment by cpuhrsch** [comment]
I'm out of depth for this particular code-piece and will leave the review to @fmassa

**Comment by rronan** [comment]
Thanks for this feature! 

> This is how `target` should look like:
> 
> ```
> target = {}
> target["boxes"] = torch.zeros((0, 4), dtype=torch.float32)
> target["labels"] = torch.zeros((1, 1), dtype=torch.int64)
> target["image_id"] = image_id
> target["area"] = (boxes[:, 3] - boxes[:, 1]) * (boxes[:, 2] - boxes[:, 0])
> target["iscrowd"] = torch.zeros((0,), dtype=torch.int64)
> ```

One question though: why does target["labels"] look like this? Shouldn't it be of length 0 (like boxes), or with only one dimension (not two) like specified here:
https://github.com/pytorch/vision/blob/master/torchvision/models/detection/faster_rcnn.py#L47

**Comment by fmassa** [comment]
@rronan I agree, I was wondering the same thing. @mnc537 is this a typo in the test?

**Comment by mnc537** [comment]
@rronan, right! It should be of length 0. There is a typo in the test, @fmassa. I'll fix it.

**Comment by fmassa** [comment]
Thanks @mnc537 !

**Comment by rronan** [comment]
Thank you @mnc537. I've been using this PR and did not experience any issue yet.

**Comment by Kirayue** [comment]
Hi, @fmassa, @mnc537 

I have a question about the targe["labels"],

According to the PR, 
```
target = {}
target["boxes"] = torch.zeros((0, 4), dtype=torch.float32)
target["labels"] = torch.zeros((1, 1), dtype=torch.int64)
target["image_id"] = image_id
target["area"] = (boxes[:, 3] - boxes[:, 1]) * (boxes[:, 2] - boxes[:, 0])
target["iscrowd"] = torch.zeros((0,), dtype=torch.int64)
```
@mnc537 said there was a typo, so the target should look like
`target["labels"] = torch.zeros((1, ), dtype=torch.int64)` or 
`target["labels"] = torch.ones((0, ), dtype=torch.int64)` or 
`target["labels"] = torch.zeros((1, 1), dtype=torch.int64)`.

I tried all above but there weren't any error.

Thank you for the feature.

**Comment by dccf36** [comment]
Does anyone encounter the problem that loss explodes after I add the negative background

**Comment by samra-irshad** [comment]
@mnc537 Thanks for this feature. Just wondering, do we need to increase the number of classes once we add the targets for negative samples?

**Comment by Kirayue** [comment]
> @mnc537 Thanks for this feature. Just wondering, do we need to increase the number of classes once we add the targets for negative samples?

The number of classes is the number of your object + 1(negative samples, or image with only background).

image

https://pytorch.org/vision/stable/models.html#id35

**Comment by samra-irshad** [comment]
@Kirayue So background class (0) and negative samples (images with no object) should have same label? Or I should allocate an additional label to images with no objects?

**Comment by Kirayue** [comment]
@samra-irshad, they are the same, so you do not need to add an additional label to indicate no objects.

**Comment by truong2710-cyber** [comment]
I have changed my code for negative images to :
target = {}
target["boxes"] = torch.zeros((0, 4), dtype=torch.float32)
target["labels"] = torch.zeros((1, ), dtype=torch.int64)
target["image_id"] = image_id
target["area"] = (boxes[:, 3] - boxes[:, 1]) * (boxes[:, 2] - boxes[:, 0])
target["iscrowd"] = torch.zeros((0,), dtype=torch.int64)

but after training I got negative values of precision and recall. Can anyone help me?
image
